src/components/Form/types.ts

```diff
@@ -1,58 +1,71 @@
+type ValidInputs =
+    | typeof TextInput
+    | typeof AmountTextInput
+    | typeof SingleChoiceQuestion
+    | typeof CheckboxWithLabel
+    | typeof Picker
+    | typeof AddressSearch
+    | typeof CountrySelector
+    | typeof StatePicker;
 
 type ValueTypeKey = 'string' | 'boolean' | 'date';
+type ValueTypeMap = {
+    string: string;
+    boolean: boolean;
+    date: Date;
+};
+type FormValue = ValueOf<ValueTypeMap>;
 
-type MeasureLayoutOnSuccessCallback = (left: number, top: number, width: number, height: number) => void;
+type InputComponentValueProps<TValue extends ValueTypeKey = ValueTypeKey> = {
+    valueType?: TValue;
+    value?: ValueTypeMap[TValue];
+    defaultValue?: ValueTypeMap[TValue];
+    onValueChange?: (value: ValueTypeMap[TValue], key: string) => void;
+    shouldSaveDraft?: boolean;
+    shouldUseDefaultValue?: boolean;
+};
 
-type BaseInputProps = {
+type MeasureLayoutOnSuccessCallback = (left: number, top: number, width: number, height: number) => void;
+type InputComponentBaseProps = InputComponentValueProps & {
+    inputID?: string;
+    errorText?: string;
     shouldSetTouchedOnBlurOnly?: boolean;
-    onValueChange?: (value: unknown, key: string) => void;
+    isFocused?: boolean;
+    measureLayout?: (ref: unknown, callback: MeasureLayoutOnSuccessCallback) => void;
+    focus?: () => void;
     onTouched?: (event: GestureResponderEvent) => void;
-    valueType?: ValueTypeKey;
-    value?: FormValueType;
-    defaultValue?: FormValueType;
     onBlur?: (event: FocusEvent | NativeSyntheticEvent<TextInputFocusEventData>) => void;
     onPressOut?: (event: GestureResponderEvent) => void;
     onPress?: (event: GestureResponderEvent) => void;
-    shouldSaveDraft?: boolean;
-    shouldUseDefaultValue?: boolean;
-    key?: Key | null | undefined;
+    onInputChange?: (value: FormValue, key: string) => void;
+    key?: Key;
     ref?: Ref<unknown>;
-    isFocused?: boolean;
-    measureLayout?: (ref: unknown, callback: MeasureLayoutOnSuccessCallback) => void;
-    focus?: () => void;
 };
 
-type InputWrapperProps<TInput extends ValidInputs> = Omit<BaseInputProps, 'ref'> &
-    ComponentProps<TInput> & {
-        InputComponent: TInput;
-        inputID: string;
-    };
-
-type ExcludeDraft<T> = T extends `${string}Draft` ? never : T;
-type OnyxFormKeyWithoutDraft = ExcludeDraft<OnyxFormKey>;
-
-type OnyxFormValues<TOnyxKey extends OnyxFormKey & keyof OnyxValues = OnyxFormKey> = OnyxValues[TOnyxKey];
-type OnyxFormValuesFields<TOnyxKey extends OnyxFormKey & keyof OnyxValues = OnyxFormKey> = Omit<OnyxFormValues<TOnyxKey>, keyof BaseForm>;
+type FormOnyxValues<TFormID extends OnyxFormKey = OnyxFormKey> = Omit<OnyxValues[TFormID], keyof BaseForm>;
+type FormOnyxKeys<TFormID extends OnyxFormKey = OnyxFormKey> = FormOnyxValues<TFormID> extends EmptyObject ? string : keyof FormOnyxValues<TFormID>;
```

Comment:
> When would FormOnyxValues<TFormID> extend EmptyObject?

When no inputs are defined for a form, like in `REPORT_FIELD_EDIT_FORM`:
```ts
[ONYXKEYS.FORMS.REPORT_FIELD_EDIT_FORM]: {},
```

This is because this form has a generic input wrapper with an `inputID` as any string:
```ts
<FormProvider
    style={[styles.flexGrow1, styles.ph5]}
    formID={ONYXKEYS.FORMS.REPORT_FIELD_EDIT_FORM}
    onSubmit={onSubmit}
    validate={validate}
    submitButtonText={translate('common.save')}
    enabledWhenOffline
>
    <View style={styles.mb4}>
        <InputWrapper
            InputComponent={TextInput}
            inputID={fieldID}
            name={fieldID}
            defaultValue={fieldValue}
            label={fieldName}
            accessibilityLabel={fieldName}
            role={CONST.ROLE.PRESENTATION}
            ref={inputRef}
        />
    </View>
</FormProvider>
```